File: M3Tools/Components/Database/UserDatabase.vb
Imports System.ComponentModel
Imports SPPBC.M3Tools.Types

Namespace Database
	Public NotInheritable Class UserDatabase
		Private Const path As String = "users"

		<Description("The username to use for the database connection")>
		<SettingsBindable(True)>
		Public Property Username As String
			Get
				Return dbConnection.Username
			End Get
			Set(value As String)
				dbConnection.Username = value
			End Set
		End Property

		'The password to use for the database connection
		<PasswordPropertyText(True)>
		<SettingsBindable(True)>
		<Description("The password to use for the database connection")>
		Public Property Password As String
			Get
				Return dbConnection.Password
			End Get
			Set(value As String)
				dbConnection.Password = value
			End Set
		End Property

		'The initial catalog to use for the database connection
		<Bindable(True)>
		<Description("The initial catalog to use for the database connection")>
		<SettingsBindable(True)>
		Public Property BaseUrl As String
			Get
				Return dbConnection.BaseUrl
			End Get
			Set(value As String)
				dbConnection.BaseUrl = value
			End Set
		End Property

		Public Sub CreateUser(fName As String, lName As String, email As String, username As String, password As String, Optional role As AccountRole = AccountRole.User)
			CreateUser(New User With {
				.FirstName = fName,
				.LastName = lName,
				.Email = email,
				.Login = New Auth(username, password) With {
					.Role = role
				}
			})
		End Sub

		Public Sub CreateUser(user As User)
			dbConnection.Consume(Method.Post, $"/{path}", JSON.ConvertToJSON(user))
		End Sub

		Public Function ChangePassword(username As String, oldPassword As String, newPassword As String) As Boolean
			Throw New NotImplementedException("ChangePassword")
		End Function

		Public Sub CloseAccount(userID As Integer)
			dbConnection.Consume(Method.Delete, $"/{path}/{userID}")
		End Sub

		Public Function Login(username As String, password As String) As User
			'Return Login($"{{username: '{username}', password: '{password.ToBase64String()}'}}")
			Return Login(New Auth(username, password))
		End Function

		Public Function Login(auth As String) As User
			Return dbConnection.Consume(Of User)(Method.Post, $"/{path}/login", auth).Result
		End Function

		''' <summary>
		''' Attempt to login a user provided their username and password
		''' </summary>
		''' <param name="auth">The credentials to use for logging in the user</param>
		''' <returns>The user if successful, otherwise Nothing</returns>
		Public Function Login(auth As Auth) As User
			Return dbConnection.Consume(Of User)(Method.Post, $"/{path}/login", JSON.ConvertToJSON(auth)).Result
		End Function

		Public Function GetUser(userID As Integer) As User
			Return dbConnection.Consume(Of User)(Method.Get, $"/{path}/{userID}").Result
		End Function

		Function GetUsers() As DBEntryCollection(Of User)
			Return dbConnection.Consume(Of DBEntryCollection(Of User))(Method.Get, $"/{path}").Result
		End Function

		'Private Structure ColumnNames
		'	Shared ReadOnly Property ID As String = "UserID"
		'	Shared ReadOnly Property FirstName As String = "FirstName"
		'	Shared ReadOnly Property LastName As String = "LastName"
		'	Shared ReadOnly Property Email As String = "Email"
		'	Shared ReadOnly Property Username As String = "Username"
		'	Shared ReadOnly Property Password As String = "Password"
		'	Shared ReadOnly Property Role As String = "AccountRole"
		'	Shared ReadOnly Property Salt As String = "Salt"
		'	Shared ReadOnly Property LastLogin As String = "LastLogin"
		'	Shared ReadOnly Property Joined As String = "Joined"

		'	Shared ReadOnly Property Message As String = "Message"
		'End Structure
	End Class
End Namespace

File: M3Tools/Components/Database/UserDatabase.cs
﻿using System;
using System.Diagnostics;
using System.Text.RegularExpressions;
using System.Threading.Tasks;

using SPPBC.M3Tools.M3API;

namespace SPPBC.M3Tools.API
{
	/// <summary>
	/// 
	/// </summary>
	public sealed partial class UserDatabase : ApiBase
	{
		/// <summary>
		/// 
		/// </summary>
		public UserDatabase() : base() => InitializeComponent();

		/// <summary>
		/// Create a new user for the application
		/// </summary>
		/// <param name="user"></param>
		/// <param name="ct"></param>
		public async void CreateUser(Types.User user, System.Threading.CancellationToken ct = default)
			=> await ExecuteAsync(System.Net.Http.HttpMethod.Post, Paths.Users, JSON.ConvertToJSON(user), ct);

		/// <summary>
		/// Change the password of the specified user
		/// </summary>
		/// <param name="username"></param>
		/// <param name="oldPassword"></param>
		/// <param name="newPassword"></param>
		/// <param name="ct"></param>
		/// <returns></returns>
		/// <exception cref="NotImplementedException"></exception>
		public bool ChangePassword(string username, string oldPassword, string newPassword, System.Threading.CancellationToken ct = default)
			=> throw new NotImplementedException("ChangePassword");

		/// <summary>
		/// Completely delete the specified user's account
		/// </summary>
		/// <param name="userID"></param>
		/// <param name="ct"></param>
		public async void CloseAccount(int userID, System.Threading.CancellationToken ct = default)
			=> await ExecuteAsync(System.Net.Http.HttpMethod.Delete, string.Join("/", Paths.Users, userID)/*$"{Paths.Users}/{userID}"*/, string.Empty, ct);

		/// <summary>
		/// Login a user using the provided login info
		/// </summary>
		/// <param name="username"></param>
		/// <param name="password"></param>
		/// <param name="ct"></param>
		/// <returns></returns>
		public async Task<Types.User> Login(string username, string password, System.Threading.CancellationToken ct = default)
			=> await Login(new Types.Auth(username, password), ct);

		/// <summary>
		/// Attempt to login a user provided their username and password
		/// </summary>
		/// <param name="auth">The credentials to use for logging in the user</param>
		/// <param name="ct"></param>
		/// <returns>The user if successful, otherwise Nothing</returns>
		private async Task<Types.User> Login(Types.Auth auth, System.Threading.CancellationToken ct)
		{
			try
			{
				Debug.WriteLine($"Logging in user with username {auth.Username}...");
				return ParseResponse<Types.User>(await ExecuteWithResultAsync(System.Net.Http.HttpMethod.Post, $"{Paths.Users}/login", JSON.ConvertToJSON(auth), ct));
			}
			catch (Exception ex)
			{
				Exception baseEx = ex.GetBaseException();

				throw baseEx.Message switch
				{
					var username when Regex.IsMatch(username, @".*username.*", RegexOptions.IgnoreCase) => new Exceptions.UsernameException(),
					var password when Regex.IsMatch(password, @".*password.*", RegexOptions.IgnoreCase) => new Exceptions.PasswordException(),
					_ => baseEx
				};
			}
		}
	}
}
